### PR TITLE
feat: Support 3.12 python

### PR DESCRIPTION
**Pull request by bcm-at-zama**
Adding support for 3.12

closes #https://github.com/zama-ai/concrete-ml-internal/issues/4083

### CODE REVIEW

**Comment by bcm-at-zama** [comment]
CP 3.11 installation was fixed: it was a bad command line.

**Comment by bcm-at-zama** [comment]
Looks good to me, but I haven't ran it on 3.12.

**Comment by bcm-at-zama** [comment]
Added a reversible commit to try 3.12 on the PR

**Comment by bcm-at-zama** [comment]
We have an issue in `make sync_env`:

```
 • Installing websocket-client (1.8.0)

  ChefBuildError

  Backend 'setuptools.build_meta:__legacy__' is not available.
  
  Cannot import 'setuptools.build_meta'

  at /opt/hostedtoolcache/Python/3.12.5/x64/lib/python3.12/site-packages/poetry/installation/chef.py:164 in _prepare
      160│ 
      161│                 error = ChefBuildError("\n\n".join(message_parts))
      162│ 
      163│             if error is not None:
    → 164│                 raise error from None
      165│ 
      166│             return path
      167│ 
      168│     def _prepare_sdist(self, archive: Path, destination: Path | None = None) -> Path:

Note: This error originates from the build backend, and is likely not a problem with poetry but with numpy (1.23.5) not supporting PEP 517 builds. You can verify this by running 'pip wheel --no-cache-dir --use-pep517 "numpy (==1.23.5)"'.

make: *** [Makefile:32: setup_env] Error 1
```

**Comment by bcm-at-zama** [comment]
Trying on my machine

**Comment by bcm-at-zama** [comment]
Trying to see if I can change a bit the toml and make sync_env with 3.12

**Comment by bcm-at-zama** [comment]
@andrei-stoian-zama : do I let you finish this PR?

File: pyproject.toml [hunk 1]
```diff
@@ -59,16 +59,25 @@ onnxoptimizer = "0.3.13"
...
+        {version = "1.26.4", python = ">=3.9.0"}        
+        ]
+protobuf = "3.20.3"
 pandas = "2.0.3"
 
 # Zama sources
 [[tool.poetry.source]]
 name = "zama-pypi-cpu"
-url = "https://pypi.zama.ai/cpu"
+url = "https://pypi.zama.ai/gpu"
```
Comment (andrei-stoian-zama) [inline pyproject.toml:80]: don't forget to switch back

File: pyproject.toml [hunk 2]
```diff
@@ -59,16 +59,25 @@ onnxoptimizer = "0.3.13"
...
+        ]
+scipy = [
+        {version = "1.10.1", python = "3.8"},
+        {version = ">=1.10.1", python = ">3.8"}
+        ]
+numpy = [
+        {version = "1.23.5", python = "<3.9" }, 
+        {version = "1.26.4", python = ">=3.9.0"}        
+        ]
+protobuf = "3.20.3"
```
Comment (andrei-stoian-zama) [inline pyproject.toml:74]: I'm not sure why we need these specific versions now.. didn't we need them already when supporting python3.11 ?

File: script/make_utils/setup_os_deps.sh [hunk 3]
```diff
@@ -136,10 +136,8 @@ if [[ "${OS_NAME}" == "Linux" ]]; then
         openssl \
         shellcheck \
         texlive-latex-base texlive-latex-extra texlive-fonts-recommended texlive-xetex lmodern \
-        wget && \
+        wget pipx adduser && \
```
Comment (andrei-stoian-zama) [inline script/make_utils/setup_os_deps.sh:139]: with ubuntu>23 we install poetry using `pipx`

File: docker/Dockerfile.dev [hunk 4]
```diff
@@ -38,6 +44,13 @@ RUN groupadd -g "${BUILD_GID}" dev_user && \
 # Now switch to the newly created user
 USER dev_user
 
+RUN (pipx install poetry && pipx ensurepath) || \
```
Comment (andrei-stoian-zama) [inline docker/Dockerfile.dev:47]: if installation fails with `pipx` (ubuntu <=22), install in the original manner and add the user's local `site-packages` to the path

File: docker/Dockerfile.dev [hunk 5]
```diff
@@ -1,4 +1,5 @@
-FROM ubuntu:20.04
+ARG UBUNTU_BASE=20.04
+FROM ubuntu:${UBUNTU_BASE}
```
Comment (andrei-stoian-zama) [inline docker/Dockerfile.dev:2]: can now be parametrized by the caller

File: Makefile [hunk 6]
```diff
@@ -1,9 +1,17 @@
 SHELL:=$(shell /usr/bin/env which bash)
 
+DEV_DOCKER_PYTHON?=py38
+$(warning now we reached ifeq DEV_DOCKER_PYTHON=$(DEV_DOCKER_PYTHON))
```
Comment (andrei-stoian-zama) [inline Makefile:4]: remove this warning

File: Makefile [hunk 7]
```diff
@@ -342,12 +350,13 @@ mypy_ci:
 
 .PHONY: docker_build # Build dev docker
 docker_build:
-	BUILD_ARGS=; \
+	BUILD_ARGS="--build-arg UBUNTU_BASE=$(UBUNTU_BASE) ";\
+	BUILD_ARGS+="--build-arg OS_DEPS_ARG=linux-install-python "; \
```
Comment (andrei-stoian-zama) [inline Makefile:354]: remove this `OS_DEPS_ARG`

File: .github/workflows/continuous-integration.yaml [hunk 8]
```diff
@@ -21,13 +21,13 @@ on:
           - pr
 
       linux_python_versions:
-        description: "Space separated list of python versions (3.8, 3.9, 3.10, 3.11 are supported) to launch on linux"
+        description: "Space separated list of python versions (3.8, 3.9, 3.10, 3.11, 3.12 are supported) to launch on linux"
         required: false
         type: string
-        default: "3.8"
+        default: "3.12"
```
Comment (andrei-stoian-zama) [inline .github/workflows/continuous-integration.yaml:27]: revert to 3.8 here

File: script/make_utils/setup_os_deps.sh [hunk 9]
```diff
@@ -136,10 +136,8 @@ if [[ "${OS_NAME}" == "Linux" ]]; then
...
         ${CLEAR_APT_LISTS:+$CLEAR_APT_LISTS} \
-        python3 -m pip install --no-cache-dir --upgrade pip && \
-        python3 -m pip install --no-cache-dir --ignore-installed poetry==1.7.1 && \
```
Comment (andrei-stoian-zama) [inline script/make_utils/setup_os_deps.sh:142]: we need to readd this for the CI to work

**Review by jfrery — CHANGES_REQUESTED** [review]
This is a 1h45 minutes test... So 30 min + increase. Why is that?

**Comment by jfrery** [comment]
I launched a 3.12 CML test https://github.com/zama-ai/concrete-ml/actions/runs/11940082085/job/33282105181

**Review by jfrery — CHANGES_REQUESTED** [review]
We still need the CML 3.12 test to pass -> https://github.com/zama-ai/concrete-ml/actions/runs/12032176976

File: docs/advanced_examples/GLMComparison.ipynb [hunk 10]
```diff
@@ -24,7 +24,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 1,
+   "execution_count": null,
```
Comment (kcelia) [inline docs/advanced_examples/GLMComparison.ipynb:27]: Some notebooks haven't run normally, it's worth refreshing them with the CI action

**Review by jfrery — APPROVED** [review]
thanks for the test speed up!

**Pull request merged**